packages/aws-cdk-lib/core/lib/stage.ts

```diff
@@ -210,16 +216,62 @@ export class Stage extends Construct {
    * calls will return the same assembly.
    */
   public synth(options: StageSynthesisOptions = { }): cxapi.CloudAssembly {
-    if (!this.assembly || options.force) {
+
+    let newConstructPaths = this.listAllConstructPaths(this);
+
+    // If the assembly cache is uninitiazed, run synthesize and reset construct paths cache
+    if (this.constructPathsCache.size == 0 || !this.assembly || options.force) {
       this.assembly = synthesize(this, {
         skipValidation: options.skipValidation,
         validateOnSynthesis: options.validateOnSynthesis,
       });
+      newConstructPaths = this.listAllConstructPaths(this);
+      this.constructPathsCache = newConstructPaths;
     }
 
+    // If the construct paths set has changed
+    if (!this.constructPathSetsAreEqual(this.constructPathsCache, newConstructPaths)) {
+      const errorMessage = 'The construct tree has been modified after synthesis. Only the results of the first synth() call are written to disk, and modifications done after it are ignored. Avoid construct tree mutations after synth() has been called.';
+      if (options.errorOnDuplicateSynth ?? true) {
+        throw new Error(errorMessage);
+      } else {
+        // eslint-disable-next-line no-console
+        console.error(errorMessage);
```

Comment:
So we print this as a warning. And now what? The message text implies that we are in the warning situation and is trying to say the following:

- Hey we have detected that you called synth() multiple and changed the construct tree after the first synth()
- We are going to ignoring all changes, but are not failing the execution.
- You should stop doing this, unless you really know what you do.